### PR TITLE
Feat/arborist dbgap

### PR DESCRIPTION
**Pull request by Avantol13**
### New Features
- dbgap syncing now updates resources in arborist 

### Improvements
- allow not providing a user.yaml (e.g. just a pure dbgap sync). previously you had to pass an empty user.yaml
- dbgap study to arborist resource namespace configuration

### CODE REVIEW

**Comment by PlanXCyborg** [comment]
The style in this PR agrees with [`black`](https://github.com/ambv/black). :heavy_check_mark:

This formatting comment was generated automatically by a script in [uc-cdis/wool](https://github.com/uc-cdis/wool).

File: fence/sync/sync_users.py [hunk 1]
```diff
@@ -175,6 +193,18 @@ def from_file(cls, filepath, encrypted=True, key=None, logger=None):
             # to check if they're allowed to do certain things
             policies[username] = details.get("policies", [])
 
+        # get user project mapping to arborist resources if it exists
+        project_to_resource = data.get("rbac", dict()).get(
+            "user_project_to_resource", dict()
```
Comment (Avantol13) [inline fence/sync/sync_users.py]: See https://github.com/uc-cdis/commons-users/pull/485

File: fence/sync/sync_users.py [hunk 2]
```diff
@@ -443,78 +485,20 @@ def _parse_csv(self, file_dict, sess, encrypted=True):
...
+                                self._add_dbgap_project_to_arborist(
+                                    element_dict["auth_id"]
+                                )
+
                             if username not in user_projects:
                                 user_projects[username] = {}
                             user_projects[username].update(phsid_privileges)
                         except ValueError as e:
                             self.logger.info(e)
         return user_projects, user_info
```
Comment (Avantol13) [inline fence/sync/sync_users.py]: The below code was dead, never called

File: fence/sync/sync_users.py [hunk 3]
```diff
@@ -32,6 +32,9 @@
 from fence.rbac.client import ArboristClient, ArboristError
 from fence.resources.storage import StorageManager
 
+# should end in /
+DBGAP_ARBORIST_RESOURCE_PREFIX = "/dbgap/programs/"
```
Comment (Avantol13) [inline fence/sync/sync_users.py]: depending on what we decide (whether or not we actually want `dbgap` namespace) we can just change this. at least with this code we can differentiate dbgap projects if we want to

File: fence/sync/sync_users.py [hunk 4]
```diff
@@ -190,9 +218,12 @@ def from_file(cls, filepath, encrypted=True, key=None, logger=None):
             for project in details.get("projects", {}):
                 # project may not have `resource` field
                 try:
+                    # prefer resource field
                     resource = project["resource"]
                 except KeyError:
-                    continue
+                    # if no resource or mapping, assume auth_id is resource
+                    resource = project["auth_id"]
```
Comment (rudyardrichter) [inline fence/sync/sync_users.py]: is the `auth_id` ever going to be usable as a resource?

File: fence/sync/sync_users.py (hunk 2 above)
Comment (rudyardrichter) [reply fence/sync/sync_users.py]: amused and confused how this snippet pulled a disappearing act :upside_down_face:

File: fence/sync/sync_users.py [hunk 5]
```diff
@@ -1107,60 +1158,122 @@ def _update_arborist(self, session, user_yaml):
...
+                                    overwrite=True,
+                                )
+                            except ArboristError as e:
+                                self.logger.info(
+                                    "not creating policy in arborist; {}".format(str(e))
+                                )
+                            self._created_policies.add(policy_id)
 
-                    self.arborist_client.grant_user_policy(user.username, policy_id)
+                        self.arborist_client.grant_user_policy(user.username, policy_id)
```
Comment (rudyardrichter) [inline fence/sync/sync_users.py]: this chunk is making sense now :sweat_smile: it's maybe a little confusing that there's two separate groups of `grant_user_policy`, but I'm not seeing a great way to deal with that. I know it's silly but even just moving the user_yaml policy granting to the end so it's spatially closer to this one might make this function more obvious

File: tests/dbgap_sync/test_utils.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+    result = utils._get_arborist_resources_from_paths(test_paths)
+    result = utils._knead(result)
+    result = utils._undictify_subresources(result)
+
+    expected_roots = ["orgA", "orgB", "orgC", "programs"]
+    for item in result:
+        # ensure one of each of the items in expected roots
+        assert item.get("name") in expected_roots
+        expected_roots.remove(item.get("name"))
```
[the remark this answers is not in the document]
Comment (Avantol13) [reply tests/dbgap_sync/test_utils.py]: whoops, was going to assert that list is empty at the end to ensure every one got found. I'll add that

File: tests/dbgap_sync/test_utils.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+                  subresources:
+                    - name: 'test'
+            - name: 'phs000172'
+              subresources:
+                - name: 'projects'
+                  subresources:
+                    - name: 'test'
+    """
+    useryaml = yaml.safe_load(yaml_string)
+    useryaml_resources = useryaml.get("rbac", {}).get("resources")
```
[the remark this answers is not in the document]
Comment (Avantol13) [reply tests/dbgap_sync/test_utils.py]: yeeah, I'd kinda like to do similar parsing that the sync code does to ensure that the library loads the way we expect. I guess there are other tests that do the yaml loading though... :man_shrugging: I'm leaning towards leaving it but if you have a strong opinion I can change it

File: tests/dbgap_sync/test_utils.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+            )
+
+            assert "DEV" in subresources
+            assert "projects" in _get_subresources(subresources["DEV"])
+            assert "test" in _get_subresources(
+                _get_subresources(subresources["DEV"])["projects"]
+            )
+
+        elif item.get("name") == "gen3":
+            assert "programs" in subresources
```
[the remark this answers is not in the document]
Comment (Avantol13) [reply tests/dbgap_sync/test_utils.py]: oh, good point

File: fence/sync/sync_users.py [hunk 9]
```diff
@@ -1107,60 +1158,122 @@ def _update_arborist(self, session, user_yaml):
...
+                    {"name": dbgap_study, "description": "synced from dbGaP"},
+                    create_parents=True,
+                )
+                self.logger.info(
+                    "added arborist resource under parent path: {} for dbgap project {}.".format(
+                        resource_namespace, dbgap_study
+                    )
+                )
+                self.logger.debug("Arborist response: {}".format(response))
+                if dbgap_study not in self._dbgap_study_to_resources:
```
[the remark this answers is not in the document]
Comment (Avantol13) [reply fence/sync/sync_users.py]: oh, for sure, I'll update it

File: fence/sync/sync_users.py [hunk 10]
```diff
@@ -175,6 +191,18 @@ def from_file(cls, filepath, encrypted=True, key=None, logger=None):
             # to check if they're allowed to do certain things
             policies[username] = details.get("policies", [])
 
+        # get user project mapping to arborist resources if it exists
+        project_to_resource = data.get("rbac", dict()).get(
+            "user_project_to_resource", dict()
+        )
+
+        if logger:
+            logger.info(
```
[the remark this answers is not in the document]
Comment (Avantol13) [reply fence/sync/sync_users.py]: the logger is optionally passed in and this is a `classmethod` so no access to `self`. I made the same mistake the first time adding this

File: fence/sync/sync_users.py (hunk 4 above)
Comment (Avantol13) [reply fence/sync/sync_users.py]: what do you mean? like is the `Project.auth_id` ever going to point to an arborist resource path?

File: fence/sync/sync_users.py (hunk 5 above)
Comment (Avantol13) [reply fence/sync/sync_users.py]: gotcha, that's an easy change :smile: I'll move it towards the bottom

File: fence/sync/sync_users.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (rudyardrichter) [reply fence/sync/sync_users.py]: ooh, sorry. hard to follow that on github

File: fence/sync/sync_users.py (hunk 4 above)
Comment (rudyardrichter) [reply fence/sync/sync_users.py]: yeah like `Project.auth_id` will be stuff like `DEV-test`, right? so not relevant to arborist?
